File: src/main/java/toby/jpa/persistence/impl/BrotherPersistenceImpl.java
package toby.jpa.persistence.impl;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.Query;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;
import toby.jpa.dto.BrotherDto;
import toby.jpa.persistence.IBrotherPersistence;

import java.util.List;

@Repository
@Transactional
public class BrotherPersistenceImpl implements IBrotherPersistence  {

@PersistenceContext
protected EntityManager em;

public EntityManager getEntityManager(){
        return em;
        }

public void setEntityManager(EntityManager entityManager){
        this.em=entityManager;
        }


@Override
public BrotherDto getBrotherById(Long discordId){
        return em.find(BrotherDto.class,discordId);
        }

@Override
public BrotherDto getUserByName(String name){
        Query q=em.createNamedQuery("BrotherDto.getName",BrotherDto.class);
        q.setParameter("name",name);
        return(BrotherDto)q.getSingleResult();
        }

@Override
public BrotherDto updateBrother(BrotherDto brotherDto){
        em.merge(brotherDto);
        em.flush();
        return brotherDto;
        }

@Override
@SuppressWarnings("unchecked")
public List<BrotherDto> listBrothers(){
        Query q=em.createNamedQuery("BrotherDto.getAll",BrotherDto.class);
        return q.getResultList();
        }


@Override
public BrotherDto createNewBrother(BrotherDto brotherDto){

        BrotherDto databaseBrother=em.find(BrotherDto.class,brotherDto.getDiscordId());
        if(databaseBrother==null){
        return persistBrotherDto(brotherDto);
        }else if(!brotherDto.getDiscordId().equals(databaseBrother.getDiscordId())){
        return persistBrotherDto(brotherDto);
        }else
        return databaseBrother;
        }

@Override
public void deleteBrother(BrotherDto brotherDto){
        em.remove(brotherDto);
        em.flush();
        }

@Override
public void deleteBrotherById(Long discordId){
        Query q=em.createNamedQuery("BrotherDto.deleteById");
        q.setParameter("discordId",discordId);
        q.executeUpdate();
        }

private BrotherDto persistBrotherDto(BrotherDto brotherDto){
        em.persist(brotherDto);
        em.flush();
        return brotherDto;
        }
        }

File: database/src/main/kotlin/database/persistence/impl/BrotherPersistenceImpl.kt
package database.persistence.impl

import database.persistence.IBrotherPersistence
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import jakarta.persistence.Query
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional
open class BrotherPersistenceImpl : IBrotherPersistence {
    @PersistenceContext
    lateinit var entityManager: EntityManager


    override fun getBrotherById(discordId: Long?): database.dto.BrotherDto {
        return entityManager.find(database.dto.BrotherDto::class.java, discordId)
    }

    override fun getUserByName(name: String?): database.dto.BrotherDto {
        val q: Query = entityManager.createNamedQuery("BrotherDto.getName", database.dto.BrotherDto::class.java)
        q.setParameter("name", name)
        return q.singleResult as database.dto.BrotherDto
    }

    override fun updateBrother(brotherDto: database.dto.BrotherDto?): database.dto.BrotherDto? {
        entityManager.merge(brotherDto)
        entityManager.flush()
        return brotherDto
    }

    override fun listBrothers(): List<database.dto.BrotherDto?> {
        val q: Query = entityManager.createNamedQuery("BrotherDto.getAll", database.dto.BrotherDto::class.java)
        return q.resultList as List<database.dto.BrotherDto?>
    }


    override fun createNewBrother(brotherDto: database.dto.BrotherDto): database.dto.BrotherDto {
        val databaseBrother = entityManager.find(database.dto.BrotherDto::class.java, brotherDto.discordId)
        return if (databaseBrother == null) {
            persistBrotherDto(brotherDto)
        } else if (brotherDto.discordId != databaseBrother.discordId) {
            persistBrotherDto(brotherDto)
        } else databaseBrother
    }

    override fun deleteBrother(brotherDto: database.dto.BrotherDto?) {
        entityManager.remove(brotherDto)
        entityManager.flush()
    }

    override fun deleteBrotherById(discordId: Long?) {
        val q = entityManager.createNamedQuery("BrotherDto.deleteById")
        q.setParameter("discordId", discordId)
        q.executeUpdate()
    }

    private fun persistBrotherDto(brotherDto: database.dto.BrotherDto): database.dto.BrotherDto {
        entityManager.persist(brotherDto)
        entityManager.flush()
        return brotherDto
    }
}
